src/coreclr/debug/runtimeinfo/CMakeLists.txt

```diff
@@ -37,3 +37,46 @@ endif()
 
 # publish runtimeinfo lib
 install_clr(TARGETS runtimeinfo DESTINATIONS lib COMPONENT runtime)
+
+add_library(cdac_data_descriptor OBJECT datadescriptor.cpp)
+# don't build the data descriptor before the VM (and any of its dependencies' generated headers)
+add_dependencies(cdac_data_descriptor cee_wks_core)
+if(CLR_CMAKE_TARGET_WIN32)
+  # turn off whole program optimization:
+  # 1. it creates object files that cdac-build-tool can't read
+  # 2. we never link cdac_data_descriptor into the final product - it's only job is to be scraped
+  target_compile_options(cdac_data_descriptor PRIVATE /GL-)
+endif()
+target_include_directories(cdac_data_descriptor BEFORE PRIVATE ${VM_DIR})
+target_include_directories(cdac_data_descriptor BEFORE PRIVATE ${VM_DIR}/${ARCH_SOURCES_DIR})
+target_include_directories(cdac_data_descriptor PRIVATE ${CLR_DIR}/interop/inc)
+
+set(GENERATED_CDAC_DESCRIPTOR_DIR "${CMAKE_CURRENT_BINARY_DIR}/cdac")
+set(CONTRACT_DESCRIPTOR_OUTPUT "${GENERATED_CDAC_DESCRIPTOR_DIR}/contract-descriptor.c")
+if("${CDAC_BUILD_TOOL_BINARY_PATH}" STREQUAL "" OR NOT EXISTS "${CDAC_BUILD_TOOL_BINARY_PATH}")
+  message(FATAL_ERROR "No cdac-build-tool set or ${CDAC_BUILD_TOOL_BINARY_PATH} does not exist")
+endif()
+
+set(CONTRACT_FILE "${CMAKE_CURRENT_SOURCE_DIR}/contracts.jsonc")
+
+# generate the contract descriptor by running cdac-build-tool
+# n.b. this just uses `dotnet` from the PATH.  InitializeDotNetCli adds the apropropriate directory
+add_custom_command(
+    OUTPUT "${CONTRACT_DESCRIPTOR_OUTPUT}"
+    VERBATIM
+    COMMAND dotnet ${CDAC_BUILD_TOOL_BINARY_PATH} compose -o "${CONTRACT_DESCRIPTOR_OUTPUT}" -c "${CONTRACT_FILE}" $<TARGET_OBJECTS:cdac_data_descriptor>
```

Comment:
1. `src/coreclr/build-runtime.sh` direct invocation skips requiring dotnet/msbuild, which is necessary for new platform port work. This PR assumes dotnet/msbuild are always available. We should handle that case.

2. Secondly, can use published cdac-build-tool instead of assembly so instead of requiring `dotnet` in PATH, it uses executable (AOT'd, singilefilehost, apphost -based doesn't matter)? If that's not possible, then please pass `DOTNET_HOST_PATH` (set by SDK to the executing msbuild) via cmakeargs as well.

Reply:
Can we just make embedding the contract descriptor optional if someone is doing a build without msbuild?

Reply:
> If making it optional doesn't affect debugging, that's good

in the future, it will make using something like dotnet-sos impossible

I don't understand what scenarios don't have a .NET sdk available.  any new platform bringup will start with cross compiling from an existing platform with a .NET sdk and a cross compiler toolchain, right?  that case should be handled by the cmake+msbuild infrastructure


I added https://github.com/dotnet/runtime/pull/101297

---

> ```diff
> -    COMMAND dotnet ${CDAC_BUILD_TOOL_BINARY_PATH} compose -o "${CONTRACT_DESCRIPTOR_OUTPUT}" -c "${CONTRACT_FILE}" $<TARGET_OBJECTS:cdac_data_descriptor>
> +    COMMAND "${CLR_DOTNET_HOST_PATH}" "${CDAC_BUILD_TOOL_BINARY_PATH}" compose -o "${CONTRACT_DESCRIPTOR_OUTPUT}" -c "${CONTRACT_FILE}" $<TARGET_OBJECTS:cdac_data_descriptor>
> ```

I had something like this before and it broke source build because it was pointing to the repo root `.dotnet/dotnet`, not to `artifacts/sb/.dotnet/dotnet`.  AFAICT it's fine to just call "dotnet" from the path.  eng/commons/build.sh alaways adds the right SDK dir to the path - it's done by InitializeDotNetCli, and it is source-build aware

Reply:
*Update* I'm educating myself over in https://github.com/dotnet/installer/pull/19534#issuecomment-2066812161

I think previously I used something other than `DOTNET_HOST_PATH`

Reply:
Updated https://github.com/dotnet/runtime/pull/101297 to use DOTNET_HOST_PATH if it's set